framework/spec.go

```diff
@@ -1,85 +0,0 @@
-	// custom object currently known to the informer. On update events the
-	// informer knows about the old and the new custom object.
-	// GetDesiredState then receives the new custom object to be able to
-	// compute the desired state of a system.
-	GetDesiredState(ctx context.Context, obj interface{}) (interface{}, error)
-
-	// NewUpdatePatch is called upon observed custom object change. It receives
-	// the observed custom object, the current state as provided by
-	// GetCurrentState and the desired state as provided by
-	// GetDesiredState. NewUpdatePatch analyses the current and desired
-	// state and returns the patch to be applied by Create, Delete, and
-	// Update functions. ApplyCreateChange, ApplyDeleteChange, and
-	// ApplyUpdateChange are called only when the corresponding patch part
-	// was created.
-	NewUpdatePatch(ctx context.Context, obj, currentState, desiredState interface{}) (*Patch, error)
-	// NewDeletePatch is called upon observed custom object deletion. It
-	// receives the deleted custom object, the current state as provided by
-	// GetCurrentState and the desired state as provided by
-	// GetDesiredState. NewDeletePatch analyses the current and desired
-	// state returns the patch to be applied by Create, Delete, and Update
-	// functions. ApplyCreateChange, ApplyDeleteChange, and
-	// ApplyUpdateChange are called only when the corresponding patch part
-	// was created.
-	NewDeletePatch(ctx context.Context, obj, currentState, desiredState interface{}) (*Patch, error)
-
-	// ApplyCreateChange receives the new custom object observed during
-	// custom resource watches. It also receives the create portion of the
-	// Patch provided by NewUpdatePatch or NewDeletePatch.
-	// ApplyCreateChange only has to create resources based on its provided
-	// input. All other reconciliation logic and state transformation is
-	// already done at this point of the reconciliation loop.
-	ApplyCreateChange(ctx context.Context, obj, createChange interface{}) error
-	// ApplyDeleteChange receives the new custom object observed during
-	// custom resource watches. It also receives the delete portion of the
-	// Patch provided by NewUpdatePatch or NewDeletePatch.
-	// ApplyDeleteChange only has to delete resources based on its provided
-	// input. All other reconciliation logic and state transformation is
-	// already done at this point of the reconciliation loop.
-	ApplyDeleteChange(ctx context.Context, obj, deleteChange interface{}) error
-	// ApplyUpdateChange receives the new custom object observed during
-	// custom resource watches. It also receives the update portion of the
-	// Patch provided by NewUpdatePatch or NewDeletePatch.
-	// ApplyUpdateChange has to update resources based on its provided
-	// input. All other reconciliation logic and state transformation is
-	// already done at this point of the reconciliation loop.
-	ApplyUpdateChange(ctx context.Context, obj, updateChange interface{}) error
-}
```

Comment:
Bye! Replaced by crud_resource.go and crud_resource_ops.go

Reply:
So, now the methods in `CRUDResourceOps` need still to be implemented for having the full functionality, right? So, with these changes here, there are two more methods to be implemented? Or one operator could just implement the three methods in the `Resource` interface and opt-out to CRUD?

Reply:
No. There is `CRUDResource` which is an adapter. It takes `CRUDResourceOps` and transforms it to a `Resource`. See https://github.com/giantswarm/operatorkit/blob/master/framework/crud_resource.go

So for CRUD stuff you can use `CRUDResource` and implement `CRUDResourceOps` which is basically former `Resource`. Where it's an overkill or it simply doesn't make sense you implement use `Resource`.

Note that `retryresource` and `metricsresource` work with both of them.